[MRG + 1] Added verbose flag to GMM

ssaeger opened the pull request:
This PR addresses issue #4633.
Example outputs for a GMM with `n_init=2`:
`verbose = 1`

```
Expectation-maximization algorithm started.
Initialization 1
    EM iteration 1
    EM iteration 2
    EM iteration 3
    EM iteration 4
    EM iteration 5
    EM iteration 6
        EM algorithm converged.
Initialization 2
    EM iteration 1
    EM iteration 2
    EM iteration 3
    EM iteration 4
    EM iteration 5
    EM iteration 6
        EM algorithm converged.
```

`verbose > 1`

```
Expectation-maximization algorithm started.
Initialization 1
    Means have been initialized.
    Weights have been initialized.
    Covariance matrices have been initialized.
    EM iteration 1
        EM iteration 1 took 0.00023s
    EM iteration 2
        Change: 0.254791525352
        EM iteration 2 took 0.00023s
    EM iteration 3
        Change: 0.668347041104
        EM iteration 3 took 0.00023s
    EM iteration 4
        Change: 4.01160843517
        EM iteration 4 took 0.00024s
    EM iteration 5
        Change: 0.00821640659597
        EM iteration 5 took 0.00024s
    EM iteration 6
        Change: 0.0
        EM algorithm converged.
    Better parameters were found.
    Initialization 1 took 0.00660s
Initialization 2
    Means have been initialized.
    Weights have been initialized.
    Covariance matrices have been initialized.
    EM iteration 1
        EM iteration 1 took 0.00025s
    EM iteration 2
        Change: 0.254791525352
        EM iteration 2 took 0.00026s
    EM iteration 3
        Change: 0.668347041104
        EM iteration 3 took 0.00023s
    EM iteration 4
        Change: 4.01160843517
        EM iteration 4 took 0.00023s
    EM iteration 5
        Change: 0.00821640659597
        EM iteration 5 took 0.00023s
    EM iteration 6
        Change: 0.0
        EM algorithm converged.
    Initialization 2 took 0.00661s
```

Commit:
Added a verbose flag to the GMM class.

Commit:
Fixed typos.

Commit:
Fixed doctests.

amueller commented:
Can you add a the verbose flag to a test so that the lines are covered (and we make sure there is no error in them)? You need to capture the output, though. git grep StringIO to see an example in another test.

Commit:
Fixed pep8 warnings.

Commit:
Added tests for the verbose flag of GMM.

Commit:
Fixed pep8 and removed unused variable.

ssaeger commented:
I added one test for each verbosity 'level' and fixed pyflakes and pep8 warnings.

Additionally, I realized that `DPGMM` and `VBGMM` also have a verbose flag, but it is a boolean. 
Maybe it would make sense to change it to use the same values (0, 1, >1) so that it could be passed to the init method of GMM?

amueller commented:
Hum, I would like to change it to an integer, but we need to make it backward compatible.
So if someone passes a boolean, this needs to be caught in the ``fit`` (not the ``__init__``), a deprecation warning needs to be raised, and it needs to be converted to an integer (but the self.verbose attribute should not be changed).

Commit:
Adapted the verbose flag of DPGMM. Boolean flag is deprecated now.

Commit:
Added tests for the adapted verbose flag of DPGMM.

ssaeger commented:
I adapted the verbose flag in `DPGMM` and `VBGMM`, added a deprecation warning and added tests similar as for `GMM` before.

amueller commented:
can you rebase on master?

amueller commented:
lgtm, thanks :)

ssaeger commented:
Rebased on master.

GaelVaroquaux commented:
I don't understand the reason behind deprecating the boolean: a boolean is trivially cast to an integer.

ssaeger commented:
My idea was that DPGMM is based on the GMM class which uses an integer verbose flag. When DPGMM uses an integer too, we can use the verbose attribute of the base class (GMM).
Furthermore, it appears more consistent to me to use the same verbose flag for the DPGMM classes and the GMM class.

GaelVaroquaux commented:
That's not the question: accepting an integer is very good. It is not in contradiction with accepting a boolean. 

Sent from my phone. Please forgive brevity and mis spelling



On May 2, 2015, 16:44, at 16:44, "Sebastian Säger" <notifications@github.com> wrote:
>My idea was that DPGMM is based on the GMM class which uses an integer
>verbose flag. When DPGMM uses an integer too, we can use the verbose
>attribute of the base class (GMM).
>Furthermore, it appears more consistent to me to use the same verbose
>flag for the DPGMM classes and the GMM class.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/scikit-learn/scikit-learn/pull/4653#issuecomment-98367694

ssaeger commented:
That is true. So we do not need to deprecating the boolean, if we keep the behaviour for 0 and `False` the same and the behaviour for 1 and `True` the same.

I will remove the deprecation and add a test for a boolean parameter.

Commit:
Removed the deprecation and added a test for a boolean flag value.

Commit:
Cleaned the code. pep8, pyflakes and added comments.

amueller commented:
Huh, I didn't realize ``boolean > 0 `` casts. @GaelVaroquaux good to merge?

GaelVaroquaux commented:
Looks good. Merged. Thanks!
